### PR TITLE
James 2694 preliminary work

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 1]
```diff
@@ -168,6 +168,21 @@ public ZipAssert containsOnlyEntriesMatching(EntryChecks... entryChecks) throws
         }
         return myself;
     }
+    public ZipAssert containsExactlyEntriesMatching(EntryChecks... entryChecks) throws Exception {
+        isNotNull();
+        List<EntryChecks> sortedEntryChecks = Arrays.stream(entryChecks)
+            .collect(Guavate.toImmutableList());
```
Comment (trantienduchn) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: you didn't sort, then the variable name is not appropriate.

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 2]
```diff
@@ -168,6 +168,21 @@ public ZipAssert containsOnlyEntriesMatching(EntryChecks... entryChecks) throws
         }
         return myself;
     }
+    public ZipAssert containsExactlyEntriesMatching(EntryChecks... entryChecks) throws Exception {
```
Comment (trantienduchn) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: this method can reuse some part of `containsOnlyEntriesMatching()`? can you refactor it?

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java [hunk 3]
```diff
@@ -246,4 +254,32 @@ void archiveShouldWriteMailboxAnnotationsInASubDirWhenTwoPresentWithTheSameName(
                 );
         }
     }
+
+    @Test
+    void entriesInArchiveShouldBeOrderedLikeMailboxWithItsAnnotationsThenMessages() throws Exception {
```
Comment (rouazana) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java]: s/Like/As/ ?

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipEntryTypeExtraFieldTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+
+    @Test
+    void getCentralDirectoryDataShouldThrowWhenNoValue() {
+        assertThatThrownBy(() -> testee.getCentralDirectoryData())
+            .isInstanceOf(RuntimeException.class);
+    }
+
+
```
Comment (rouazana) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipEntryTypeExtraFieldTest.java]: 2 cr/lf

**Comment by remk** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by trantienduchn** [comment]
test this please

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/EntryTypeExtraField.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public EntryTypeExtraField() {
+        super();
+    }
+
+    public EntryTypeExtraField(ZipEntryType entryType) {
+        super(entryType.getValue());
+    }
+
+    public Optional<ZipEntryType> getEnumValue() throws IllegalArgumentException {
```
Comment (mbaechler) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/EntryTypeExtraField.java]: there's no argument, how can it throw IllegalArgumentException?

Comment (mbaechler) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/EntryTypeExtraField.java]: should be `zipEntryType`

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryType.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    ZipEntryType(long value) {
+        this.value = value;
+    }
+
+    private static final long MAILBOX_VALUE = 1L;
+    private static final long MAILBOX_ANNOTATION_DIR_VALUE = 2L;
+    private static final long MAILBOX_ANNOTATION_VALUE = 3L;
+    private static final long MESSAGE_VALUE = 4L;
+
+    private static final Map<Long, ZipEntryType> entryByValue = ImmutableMap.of(MAILBOX_VALUE, MAILBOX,
```
Comment (mbaechler) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryType.java]: This enum is quite annoying: every time I add an entry in it, I will have to create a int value for it and add it to the Map.
TBH, the long value is useless if you have a good testsuite (ordinal is enough) and the map can be built from values of the enum.

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipEntryTypeExtraFieldTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    @Test
+    void parseFromLocalFileDataShouldThrownWhenLengthIsSmallerThan8() {
+        byte[] input = new byte[]{0, 0, 0, 0, 0, 0, 0};
+        assertThatThrownBy(() -> testee.parseFromLocalFileData(input, 0, 7))
+            .isInstanceOf(ZipException.class);
+    }
+
+    @Test
+    void parseFromLocalFileDataShouldThrownWhenLengthIsBiggerThan8() {
+        byte[] input = new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0};
```
Comment (mbaechler) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipEntryTypeExtraFieldTest.java]: how is `new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0};` different from `new byte[9]` ?

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java (hunk 2 above)
Comment (mbaechler) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: I think that assertions should be carefully documented

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryType.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+public enum ZipEntryType {
+    MAILBOX,
+    MAILBOX_ANNOTATION_DIR,
+    MAILBOX_ANNOTATION,
+    MESSAGE;
+
+    private static final Map<Integer, ZipEntryType> entryByOrdinal;
+
+    static {
+        ZipEntryType[] elements = values();
```
Comment (blackheaven) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryType.java]: Please use Guava's `Stream.zip`

**Comment by trantienduchn** [comment]
```
[72661da3f383bec0977edb44386cfa5763c48e73] [ERROR] COMPILATION ERROR : 
[72661da3f383bec0977edb44386cfa5763c48e73] [INFO] -------------------------------------------------------------
[72661da3f383bec0977edb44386cfa5763c48e73] [INFO] --- maven-remote-resources-plugin:1.5:process (process-resource-bundles) @ apache-james-backends-jpa ---
[72661da3f383bec0977edb44386cfa5763c48e73] [ERROR] error reading /root/.m2/repository/org/assertj/assertj-core/3.10.0/assertj-core-3.10.0.jar; zip file is empty
[72661da3f383bec0977edb44386cfa5763c48e73] [ERROR] error reading /root/.m2/repository/org/assertj/assertj-core/3.10.0/assertj-core-3.10.0.jar; zip file is empty
[72661da3f383bec0977edb44386cfa5763c48e73] [ERROR] /james-project/event-sourcing/event-sourcing-pojo/src/test/java/org/apache/james/eventsourcing/EventIdTest.java:[22,35] package org.assertj.core.api does not exist
```

**Comment by remk** [comment]
test this please

**Comment by remk** [comment]
test this please

**Comment by remk** [comment]
test this please

**Comment by Arsnael** [comment]
```
[72661da3f383bec0977edb44386cfa5763c48e73] docker: Error response from daemon: No such container: keys.
```
test this please

**Comment by aduprat** [comment]
Already merged

**Pull request closed without merging**